crypto/compactcert/weights.go

```diff
@@ -0,0 +1,192 @@
+	}
+
+	// in order to make the code more readable and reusable we will define the following expressions:
+	// y = signedWeight^2 + 2^(d + 2) * signedWeight + 2^2d
+	// x = 3 * 2^b * (signedWeight^2 - 2^2d)
+	// w = d * (T - 1)
+	//
+	//  numReveals * (3 * 2^b * (signedWeight^2 - 2^2d) + d * (T-1) * Y) >= ((strengthTarget) * T + numReveals * P) * Y
+	//        /\
+	//        ||
+	//        \/
+	// numReveals * (x + w * y >= ((strengthTarget) * T + numReveals * P) * y
+	y, x, w := getSubExpressions(signedWeight)
+	lhs := &big.Int{}
+	lhs.Set(w).
+		Mul(lhs, y).
+		Add(x, lhs).
+		Mul(bigInt(numOfReveals), lhs)
+
+	revealsTimesP := &big.Int{}
+	revealsTimesP.Set(bigInt(numOfReveals)).Mul(revealsTimesP, bigInt(lnProvenWeight))
+
+	rhs := &big.Int{}
+	rhs.Set(bigInt(strengthTarget))
+	rhs.Mul(rhs, bigInt(ln2IntApproximation)).
+		Add(rhs, revealsTimesP).
+		Mul(rhs, y)
+
+	if lhs.Cmp(rhs) < 0 {
+		return ErrInsufficientSingedWeight
+	}
+
+	return nil
+}
+
+// numReveals computes the number of reveals necessary to achieve the desired
+// security target. We search for small integer that will satisfy the verification
+// inequality checked by the verifyWeights function.
+// In order to make sure the number will satisfy the verifier we will use the following inequality
+//
+// numReveals > = ((strengthTarget) * T * Y / (3 * 2^b * (signedWeight^2 - 2^2d) + (d * (T - 1) - P) * Y))
+// where signedWeight/(2^d) >=1 for some integer d>=0, p = P/(2^b) >= ln(provenWeight), t = T/(2^b) >= ln(2) >= (T-1)/(2^b)
+// for some integers P,T >= 0 and b=16.
+//
+// T and b are defined in the code as the constants ln2IntApproximation and precisionBits respectively,
+// and P is set to lnProvenWeight argument.
+//
+//
+// more details can be found on the Algorand's spec
+func numReveals(signedWeight uint64, lnProvenWeight uint64, strengthTarget uint64) (uint64, error) {
+	// in order to make the code more readable and reusable we will define the following expressions:
+	// y = signedWeight^2 + 2^(d + 2) * signedWeight + 2^2d
+	// x = 3 * 2^b * (signedWeight^2 - 2^2d)
+	// w = d * (T - 1)
+	//
+	// numReveals > = ((strengthTarget) * T * Y / (3 * 2^b * (signedWeight^2 - 2^2d) + (d * (T - 1) - P) * Y))
+	//        /\
+	//        ||
+	//        \/
+	// numReveals > = ((strengthTarget) * T * y / (x + (w - P) * y))
+	y, x, w := getSubExpressions(signedWeight)
+
+	// numerator = strengthTarget * ln2IntApproximation * y
+	numerator := bigInt(strengthTarget)
+	numerator.Mul(numerator, bigInt(ln2IntApproximation)).
+		Mul(numerator, y)
+
+	// denom =  x + (w - lnProvenWeight)  * y
+	denom := w
+	denom.Sub(denom, bigInt(lnProvenWeight)).
+		Mul(denom, y).
+		Add(x, denom)
+
+	if denom.Sign() <= 0 {
+		return 0, ErrNegativeNumOfRevealsEquation
+	}
+
+	// numberReveals = (numerator / denom) + 1
+	// by adding 1 we guarantee that the return value satisfy the inequality and therefore
+	// will satisfy the verifier.
+	res := numerator.Div(numerator, denom).Uint64() + 1
```

Comment:
Please add a comment to explain why `+1` is needed here.
Is it becase: 
// + 1 to account for the decimal point value loss due to integer division